Infinite retries on failed requests

Doy-lee opened the pull request:
This adds a new thread that blocks on a queue with a list of requests that failed to be propagated to other nodes in the swarm. This happens when there's a recursive swarm request after the initial node has received the request. Failures in those requests are stored into the node that initiated the request and is retried periodically. Retries have an exponential back-off to avoid spamming and they are pruned if the request exceeds a specific age.

Most requests are allowed to be re-submitted way past their initial timestamp as long as they are signed. There are a few exceptions to this, such as delete all and expire all. These requests are still kept around and retried, but, once a retried request returns with an error the retry is dropped immediately. This applies to all requests so the retry queue is always actively managed and pruned.

For infinite retries we now serialise some state to the SQL DB. We introduce a new `runtime_state` table that stores 2 BT encoded blobs. One containing the composition of the swarms and the list of retryable requests. 

Storing the swarms lets the storage server remember what swarm it is in, and detect changes to swarms across restarts. Since it can now remember and detect when its swarm has changed it can now do a DB dump to new members that join the swarm whilst you are offline. This should improve the "correctness" of messages held by new members in the swarm in lieu of active syncing.

Storing the retryable requests lets the retries persist across restarts. This also helps improve the "correctness" of messages in the swarm in lieu of active syncing.

Doy-lee committed:
Add --skip-bootstrap to avoid contacting of seed nodes on startup (useful for private networks)

Doy-lee committed:
Respect the skip_bootstrap_ flag

Doy-lee committed:
Get rid of db getter and setter by making it public

We don't do anything in particular in the setter or getter that might
warrant having those functions (like a lock or something) so we can
remove the indirection and call the member directly.

Doy-lee committed:
WIP: Return more extensive SN data ready payload

Doy-lee committed:
Move swarm logs into derivation function, serialise SNDataReady w/ helper function

Keep the SNDataReady serialisation code compartmentalised into one
function that handles both a read and write of the data structure.

Doy-lee committed:
Timestamps in the DB are stored in millisecond precision

Doy-lee committed:
Represent pending/ready members w/ an enum

This is preliminary work to prep for persisting the swarm member's state
to disk to allow resuming from the last known state. Currently when
a storage server starts up, it assumes that it's joining a new swarm/new
members are joining its swarm and does a full message dump to those
members instead of being able to know if a swarm member is new or we
already knew about it and then, choose a synchronisation method that
is more suited for that particular scenario.

Doy-lee committed:
Remove unused std::optional header

Doy-lee committed:
Store network swarm/swarm state to disk to resume from

Restoring the swarm state means that the if the node is an active
service node, it'll remember which swarm it was in so when the storage
server gets restarted, it can correctly detect if a new SN is joining
their swarm instead of assuming all the nodes have newly joined their
swarm and consequently dump their entire SQL db to them.

Doy-lee committed:
Update outdated comments

Doy-lee committed:
Add tmp work-around for DB swarm dump on migration to DB v1

Doy-lee committed:
Infinite retries with exponential fallback

When a node initiates a recursive swarm request, the initial node awaits
the response from all other nodes before returning to the client.
Children swarm nodes that fail to receive the request are stored into
a retryable request queue to be re-attempted later.

This queue is flushed every 3s by piggybacking onto the swarm member
check function that is periodically invoked by OMQ.

Doy-lee committed:
Get rid of duplicate fnv hash, already present in serialize.h

Doy-lee committed:
Remove unused TEST_RETRY_INTERVAl/PERIOD constants

Doy-lee committed:
Rename BTSerialise to Serialise and prefix SN's SerialiseResult to be SNSerialiseResult

Doy-lee committed:
Remove unused MessageTestStatus

Doy-lee committed:
Move retryable requests into its own thread and restore pending swarm members routine

Doy-lee committed:
Revert SN data ready response, timestamp not needed

Doy-lee committed:
Serialise retryable requests to the DB

Doy-lee committed:
Match comment w/ code regarding when to add retryable request on failure

Comment states that only timed-out requests are retried. This is correct
as an error response with error code and text are stiuations that might
mean that the recipient node is not in a valid state or will ever accept
the request in which case the safe default is to not retry to that node.

It is possible in future since all possible error states are known to
handle them specifically for the command. But for now, a sane default is
to only allow retries to nodes that were offline or we failed to
communicate with.

Doy-lee committed:
Revert NEW_SWARM_MEMBER_INTERVAL to 10s

This was originally 10s and it was mistakenly changed to 30s. We've
reverted all the changes to the swarm member checks so this should go
back to its original values as retries are handled in a separate
subsystem instead of intertwined with the member checks.

Doy-lee committed:
Update the swarms serialisation to match retryable requests patterns

Move serialise result struct for retryable requests into impl file as
it's only used locally.

Doy-lee committed:
Remove unused swarm_member_check_deadline, add comments

Doy-lee committed:
Simplify swarm member status states by removing ContactDetailsReady

The `new_swarm_member` flag disambiguates between nodes that need a DB
dump vs nodes that don't sufficiently that we don't need the
intermediate contact details ready state.

Also remove unused member var.

Doy-lee committed:
Fix dblocation typo -> db_location

Doy-lee committed:
Remove RETRY_BACKOFF_COEFF from retry request on done lambda

Not sure why, it's used in the lambda itself but it does not need to be
captured unlike the other 2 constexpr variables that are being captured
and used. Alas CI is complaining about it and treating the warning as an
error so removing it in this commit.

Doy-lee committed:
Linting

Doy-lee committed:
Request a DB dump on data-ready handshake if DB is empty

Doy-lee committed:
Add 10GiB message payload to serialise test

Doy-lee committed:
Linting

Doy-lee committed:
Reduce serialisation test to 5GiB to not OOM on CI

Doy-lee committed:
Reduce serialisation test to 100MiB to not OOM on CI

Doy-lee committed:
Stop potential infinite bootstrap of DB if swarm ID doesn't change

jagerman committed:
Make both HTTPS and QUIC listen on IPv6

Upcoming Session Router usage will use IPv6 (internally) to talk to
relays, but currently relays are only listening on the IPv4 0.0.0.0 any
address.  This adds a second HTTPS and QUIC listener on the IPv6 any
address (::) so that it will be accessible over session router.

jagerman committed:
Fix IPv6 requests

- Updates HTTP server internals to properly handle IPv6 addresses,
  including updating the rate limiter to handle them.

- Switch to using a dual-stack socket with uWebSocket (HTTP) because it
  forces dual stack on, and was failing to bind with separate IPv4/IPv6
  sockets.

- Make QUIC match dual-stack (although it doesn't strictly need to) for
  the sake of consistency.

- Change internals for request handling to use IPv6 for requests and
  rate limiting, with IPv4 addresses using mapped IPv6 addresses.

jagerman committed:
version bump

jagerman committed:
Fix tests & formatting

jagerman committed:
Merge pull request #1 from jagerman/ipv6-quic-listening

Make HTTP & QUIC listen on IPv6

jagerman committed:
Fix some versions of cli11 not liking ip= in config file

jagerman committed:
fix contrib/onion-request debug build

jagerman commented on oxenss/storage/database.hpp:
```diff
@@ -32,6 +33,13 @@ enum class StoreResult {
     Full,      // Can't insert right now because the database is full.
 };
 
+inline std::atomic<int> tmp_init_db_version = 0;
```
Mutable global variables like this are, generally, not acceptable.  As a general rule, there might be a justification for doing so, but any such justification would need to be compelling and well commented.  This particular instance is unjustifiable: this is simply tracking a database property that was assigned the last time a Database was initialized; that should be in the database object, not a global variable.

jagerman commented on oxenss/server/omq.cpp:
```diff
@@ -56,6 +61,40 @@ void OMQ::handle_sn_data_ready(oxenmq::Message& message) {
+
+            log::log(
+                    logcat,
+                    level,
```
`level` here is invariant, and is only used for the initial comparison, and in `log::log` here instead of `log::debug`; this does not improve clarity.  The whole thing can be simplified (both computationally and in terms of readability) as:
```suggestion
        if (log::get_level(logcat) <= log::Level::debug) {
            std::string label;
            if (deserialised.bt.success)
                label = "rejected, bad request payload. {})"_format(deserialised.bt.error);
            else
                label = "rejected due to bad request args";

            log::debug(
                    logcat,
```

jagerman commented on oxenss/snode/service_node.cpp:
```diff
@@ -1196,4 +1732,279 @@ void ServiceNode::process_push_batch(std::string_view blob, std::string_view sen
+        // Log the current retries
+        auto now = std::chrono::steady_clock::now();
+        if (log::Level level = log::Level::debug;
+            log::get_level(logcat) <= level && retryable_requests.size()) {
```
This is another invariant `level` that obfuscates the eventual log statement (by moving the invariant log level 60 lines away from the log statement).

It also is a bit weird in that the `level` variable here is tracking the invariant rather than the value, and would seem to be much more usefully written as:
```suggestion
        if (log::Level level = log::get_level(logcat); level <= log::Level::debug && retryable_requests.size()) {
```
with changes in the following code to use `level` instead of `log::get_level(logcat)`.

Note also that every call to `get_level` involves an atomic load, and so that change vastly reduces the amount of atomic loads in the code below.

jagerman commented on oxenss/snode/service_node.cpp:
```diff
@@ -43,29 +44,426 @@ using MISSING_PUBKEY_THRESHOLD = std::ratio<3, 100>;
+    constexpr std::string_view NODES_KEY = "u";
+    assert(COMMAND_KEY < REQ_PAYLOAD_KEY);
+    assert(REQ_PAYLOAD_KEY < CREATE_TIME_KEY);
+    assert(CREATE_TIME_KEY < NODES_KEY);
```
Checks like this that can be `static_assert`s should be.  Save `assert` for invariants only testable at runtime.

jagerman commented on oxenss/common/serialize.h:
```diff
@@ -0,0 +1,27 @@
+enum class Serialise {
+    Read,
+    Write,
+};
```
This enum should not exist.

1. It is a bool in disguise.
2. It is only ever used as a run-time parameter for a *known-at-compile-time choice* as whether to make a function do A or B, where A and B are doing fundamentally different (and in fact, opposite) things.

Effectively this just seems to end up being used to hide the word "write" or "read" in a function *argument* rather than in a function *name*.

jagerman commented on oxenss/snode/service_node.cpp:
```diff
@@ -43,29 +44,426 @@ using MISSING_PUBKEY_THRESHOLD = std::ratio<3, 100>;
+}
+
+static SerialiseRetryableRequestsResult serialize_retryable_requests(
+        Serialise serialise, std::string_view read_data, std::span<RequestRetry> write_data) {
```
See my other comments about this pattern.  This is function that bifurcates based on a parameter (which again is always known at compile time), where the bifurcation makes the function do opposite things, and the return value is forced to be a struct containing all possible (and totally orthogonal) return values from the bifurcated functions.

This should be two functions.

jagerman commented on oxenss/snode/service_node.cpp:
```diff
@@ -43,29 +44,426 @@ using MISSING_PUBKEY_THRESHOLD = std::ratio<3, 100>;
+        oxenc::bt_dict_producer d;
+        d.append(VERSION_KEY, version);
+        d.append(NEED_DB_DUMP_KEY, write_data.needs_db_dump);
+        result.bt.write_payload = d.view();
```
Avoid copying the string with:
```suggestion
        result.bt.write_payload = std::move(d).str();
```

jagerman commented on oxenss/snode/service_node.cpp:
```diff
@@ -43,29 +44,426 @@ using MISSING_PUBKEY_THRESHOLD = std::ratio<3, 100>;
+        swarm_.members_ = std::move(swarm_result.swarm_members);
+        network_.swarms_ = std::move(swarm_result.network_swarms);
+        swarm_.cur_swarm_id_ = swarm_result.swarm_cur_swarm_id;
+    } else {
```
I don't understand why all of this information needs to get stored.

- `cur_swarm_id` makes sense to me: if we switched swarms while down, knowing that lets us know if we switched swarms.
- `swarm_.members_` doesn't seem useful at all, because it's going to be obliterated by the (blocking) oxend update that comes just after the constructor runs.
- similarly for `network_.swarms_`.

The commit message here mentions:

> so when the storage server gets restarted, it can correctly detect if a new SN is joining their swarm instead of assuming all the nodes have newly joined their swarm and consequently dump their entire SQL db to them.

But I'm having trouble seeing how it does and why it should actually work that way: it seems like it should be the new swarm member's responsibility to request that dump from the other swarm members.

jagerman committed:
Remove unhelpful issue template

Storage server is sufficiently back-end that the issue template for it
just gets in the way rather than being helpful.

jagerman committed:
Slim repo cloning & repo key updates

See oxen-core PRs #1883 and #1884.

jagerman committed:
Fix missing namespace on bulk message serialization

When performing bulk serializing between SNs (which happens when a new
node joins a swarm), the message namespace was not being included and so
the receiving node would end up putting all bulk filled messages into
namespace 0, where they would then be ignored.

This fixes it to include the namespace as an extra field in the
serialization and to consume it if present.

This can be introduced without worrying about breaking older clients
because existing nodes only read the first entries of the list and
ignore any trailing items, and so this will properly get the namespace
when the two sides are both upgraded without breaking.  (New clients for
now only consume the namespace if it was included).

jagerman committed:
Merge pull request #3 from jagerman/keys-and-slim-cloning

Slim repo cloning & repo key updates

jagerman committed:
bump drone to clang-19

jagerman committed:
Merge pull request #4 from jagerman/fix-msg-ns-serialization

Fix missing namespace on bulk message serialization

Doy-lee committed:
Update static zlib to use github releases URL for reliability
